sniffles/ruletrafficgenerator.py

```diff
@@ -1632,94 +1633,147 @@ def clear_globals(self):
 
     def gen_mac_addr_from_distribution(self, sip=None, dip=None,
                                        dist_file=None):
-        if len(VENDOR_MAC_DIST) <= 0:
-            if dist_file is None:
-                self.gen_random_mac_addrs(sip, dip)
+
+        global VENDOR_MAC_DIST
+
+        if not VENDOR_MAC_DIST:
+            paths = dist_file.split(":")
+            lenPaths = len(paths)
+            source = None
+            dest = None
+            if lenPaths == 1:
+                source = dest = paths[0]
+            elif lenPaths == 2:
+                if paths[0] != "?" and paths[1] != "?":
+                    source = paths[0]
+                    dest = paths[1]
+                elif paths[0] != "?":
+                    source = paths[0]
+                elif paths[1] != "?":
+                    dest = paths[1]
             else:
-                self.create_vendor_mac_dist(dist_file)
+                print("Invalid format for mac distribution file: " + dist_file)
+                sys.exit(0)
+
+            self.create_vendor_mac_dist(source, dest)
+
+        if 'src' in VENDOR_MAC_DIST and 'dest' in VENDOR_MAC_DIST:
+            option = -1
+        elif 'src' in VENDOR_MAC_DIST:
+            option = 2
+        elif 'dest' in VENDOR_MAC_DIST:
+            option = 1
+        else:
+            option = 0
+
+        if option in [0, 1, 2]:
+            self.gen_random_mac_addrs(sip, dip, option)
 
         self.test_mac_addr_exists(sip, dip)
         if not self.s_mac:
-            self.s_mac = self.get_random_octets(self.get_dist_mac_oui())
+            self.s_mac = self.get_random_octets(self.get_dist_mac_oui(True))
             self.map_mac_addr_to_ip(self.s_mac, sip)
         if not self.d_mac:
-            self.d_mac = self.get_random_octets(self.get_dist_mac_oui())
+            self.d_mac = self.get_random_octets(self.get_dist_mac_oui(False))
             self.map_mac_addr_to_ip(self.d_mac, dip)
 
-    def gen_random_mac_addrs(self, sip=None, dip=None):
+    def gen_random_mac_addrs(self, sip=None, dip=None, option=0):
+
+        # option 0: change both s_mac and d_mac
+        # option 1: change s_mac only
+        # option 2: change d_mac only
+
         global VENDOR_MAC_OUI
         self.test_mac_addr_exists(sip, dip)
-        if not self.s_mac:
+        if not self.s_mac and (option == 0 or option == 1):
             self.s_mac = \
                 self.get_random_octets(random.choice(VENDOR_MAC_OUI))
             self.map_mac_addr_to_ip(self.s_mac, sip)
-        if not self.d_mac:
+
+        if not self.d_mac and (option == 0 or option == 2):
             self.d_mac = self.get_random_octets(random.choice(VENDOR_MAC_OUI))
             self.map_mac_addr_to_ip(self.d_mac, dip)
 
     def get_d_mac(self):
         return self.d_mac
 
-    def get_dist_mac_oui(self):
-        dist_map = VENDOR_MAC_DIST.keys()
-        pick = random.randint(1, VENDOR_MAC_DIST_DOMAIN)
+    def get_dist_mac_oui(self, isSource=True):
```

Comment:
why not just pass in the string 'src' or 'dest'?  Better yet, maybe we should make 'src' and 'dest' constants rather than magic variables.  This way there is no need for a test of isSource.